### DISCUSSION TITLE
3 Band Equilizer

### DISCUSSION THREAD

**Discussion by Grassbaecker** [pschatzmann/arduino-audiokit#16] [Q&A] [answered]
Hello  Thanks for your nice work with your Audio Tools Lib
but please can you give any Example how to use and setup the 3 Band Equilizer lib
Thanks in advance  Bernhard

**Comment by pschatzmann** [comment]
I just extended it a bit. But please notice that I did not have the time to test this class. The use is pretty simple:
Just wrap the final input or output class with a Equilizer3Bands and write or read your data using this class instead of the wrapped one. There are the regular defaultConfig() and begin() methods to be called.

just look at [the documentation](https://pschatzmann.github.io/arduino-audio-tools/html/classaudio__tools_1_1_equilizer3_bands.html)

**Reply by Grassbaecker** [reply]
Sorry but all my try with begin() fail on complie with this message
request for member 'begin' in 'eq', which is of non-class type 'audio_tools::Equilizer3Bands()'

so i have no idea what to do to init an beginn to work       but alltought your libs are nice to work with my lyrat boart

down you see my code and i hope my bad english is no to bad for you

thanks in advance and hope you can help me bernhard from Austria


#include "AudioTools.h"
#include "AudioLibs/AudioKit.h"
#include "AudioFilter/Equilizer.h"

using namespace audio_tools; 

const int BUFFER_SIZE = 1024;
uint8_t Outbuffer[BUFFER_SIZE];

uint16_t sample_rate=44100;
uint16_t channels = 2;


AudioKitStream kit; // Access I2S as stream
Equilizer3Bands eq();  //hope thats right

void setup() {
  // put your setup code here, to run once:
  Serial.begin(2000000);
  AudioLogger::instance().begin(Serial, AudioLogger::Info);
  Serial.println("starting KIT...");
  auto config = kit.defaultConfig(RXTX_MODE);
  config.sample_rate = sample_rate; 
  config.channels = channels;
  config.sd_active = false;
  config.input_device = AUDIO_HAL_ADC_INPUT_LINE2;
  kit.begin(config);
  eq.begin();  // Not work
}

void loop() {
  // put your main code here, to run repeatedly:
  delay(500);
}

**Comment by pschatzmann** [comment, marked as answer]
So if you use the https://github.com/pschatzmann/arduino-audio-tools/blob/main/examples/examples-audiokit/streams-audiokit-audiokit/streams-audiokit-audiokit.ino as starting point:
Just use Equilizer3Bands eq(kit); 
And replace the eq as copy source or copy target. It should not make any difference:

StreamCopy copier(eq, kit); // copy kit with eq to kit
or
StreamCopy copier(kit, eq); // copy kit to kit with eq

to set up the eq parameters

auto eqcfg = eq.defaultConfig();
eqcfg.bits_per_sample = kit.bits_per_sample;
eqcfg.channels = kit.channels;
eqcfg.sample_rate = kit.sample_rate;
eq.begin(eqcfg);

To update the settings just change 
eqcfg. gain_high or gain_low or gain_medium

**Reply by Grassbaecker** [reply]
Equilizer3Bands eq();  //is compliable
Equilizer3Bands eq(kit);  // is not compliable  //error  no matching function for call to 'audio_tools::Equilizer3Bands::Equilizer3Bands(audio_tools::AudioKitStream&)'

but next in the setup 

auto eqcfg = eq.defaultConfig();   //error again // request for member 'defaultConfig' in 'eq', which is of non-class type 'audio_tools::Equilizer3Bands()'


so i not understand cause i found all in the heaters but nothing will do
thanks in advance bernhard

**Comment by pschatzmann** [comment]
I think you work still with an old version of the library

**Comment by Grassbaecker** [comment]
i have download them last week as the postman was bring my lyrat bord

two day i was invest to your nice FFT lib who is quckliy but i was have some proplem
for my project i was need the full magnitue arraey and i was change the magnitude function in the heater from only return the val to change first the array and after retunn the val but the arry have not changed
so i was go the long way and have read in both arrays and have self them squrerootet
Nice all fast enoug   but them now way with my Oled lib who hav now problem with now Audiotool kit and them to ArduinoFFT  who is fucking slowly with its double precision val

will try today maybe RealFFT alone without AudioFFT will work for me

thanks in advance bernhard

**Comment by pschatzmann** [comment]
You should not use doubles but stick to floats.
I just committed the change where I made the  float magnitude(idx) function public. You can easily fill your own magnitude array by looping over your array from 0 to size() calling magnitude(idx). I also added a new magnitudes() function which returns the result as array of floats but note that this is allocating additional memory!
I was comparing the speed of different FFT implementations but I did not find any difference in any of them!

ps. you will need to upgrade the library to use the new functionality!

**Comment by Grassbaecker** [comment]
i have llok into new sources and the new 
float* magnitudes() {
            if (p_magnitudes==nullptr){
                p_magnitudes = new float[size()];
            }
            for (int j=0;j<size();j++){
                p_magnitudes[j]= magnitude(j);
            }
            return p_magnitudes;

but would in not better to for (int j=0;j<size()/2;j++)    couse it is only the half of the array usfull t read

thanks in advance bernhard

**Comment by pschatzmann** [comment]
You are right: I was mixing up the meaning of size(). It was intended to provide the number of relevant bins for the result: cfg.length/2.
The correction has been committed
